src/aktualizr_secondary/secondary_rpc_test.cc

```diff
@@ -61,78 +211,190 @@ bool operator==(const Uptane::RawMetaPack& lhs, const Uptane::RawMetaPack& rhs)
+  Uptane::SecondaryInterface::Ptr ip_secondary_;
+};
+
 // Test the serialization/deserialization and the TCP/IP communication implementation
 // that occurs during communication between Primary and IP Secondary
-TEST(SecondaryTcpServer, TestIpSecondaryRPC) {
-  // secondary object on Secondary ECU
-  SecondaryMock secondary(Uptane::EcuSerial("serial"), Uptane::HardwareIdentifier("hardware-id"),
-                          PublicKey("pub-key", KeyType::kED25519), Uptane::Manifest());
-
-  // create Secondary on Secondary ECU, and run it in a dedicated thread
-  SecondaryTcpServer secondary_server(secondary.getDispatcher(), "", 0);
-  std::thread secondary_server_thread{[&secondary_server]() { secondary_server.run(); }};
-
-  secondary_server.wait_until_running();
-  // create Secondary on Primary ECU, try it a few times since the secondary thread
-  // might not be ready at the moment of the first try
-  Uptane::SecondaryInterface::Ptr ip_secondary =
-      Uptane::IpUptaneSecondary::connectAndCreate("localhost", secondary_server.port());
-
-  ASSERT_TRUE(ip_secondary != nullptr) << "Failed to create IP Secondary";
-  EXPECT_EQ(ip_secondary->getSerial(), secondary.getSerial());
-  EXPECT_EQ(ip_secondary->getHwId(), secondary.getHwId());
-  EXPECT_EQ(ip_secondary->getPublicKey(), secondary.getPublicKey());
-  EXPECT_EQ(ip_secondary->getManifest(), secondary.getManifest());
+TEST_P(SecondaryRpcTest, AllRpcCallsTest) {
+  ASSERT_TRUE(ip_secondary_ != nullptr) << "Failed to create IP Secondary";
+  EXPECT_EQ(ip_secondary_->getSerial(), secondary_.serial());
+  EXPECT_EQ(ip_secondary_->getHwId(), secondary_.hwID());
+  EXPECT_EQ(ip_secondary_->getPublicKey(), secondary_.publicKey());
+  EXPECT_EQ(ip_secondary_->getManifest(), secondary_.manifest());
 
   Uptane::RawMetaPack meta_pack{"director-root", "director-target", "image_root",
                                 "image_targets", "image_timestamp", "image_snapshot"};
 
-  EXPECT_TRUE(ip_secondary->putMetadata(meta_pack));
-  EXPECT_TRUE(meta_pack == secondary._metapack);
-
-  std::string firmware = "firmware";
-  EXPECT_TRUE(ip_secondary->sendFirmware(firmware));
-  EXPECT_EQ(firmware, secondary._data);
+  EXPECT_TRUE(ip_secondary_->putMetadata(meta_pack));
+  EXPECT_TRUE(meta_pack == secondary_.metadata());
 
-  EXPECT_EQ(ip_secondary->install(""), data::ResultCode::Numeric::kOk);
+  EXPECT_EQ(sendAndInstallBinaryImage(), data::ResultCode::Numeric::kOk);
+  EXPECT_EQ(image_file_.hash(), secondary_.getReceivedImageHash());
 
-  secondary_server.stop();
-  secondary_server_thread.join();
+  EXPECT_EQ(installOstreeRev(), data::ResultCode::Numeric::kOk);
+  EXPECT_EQ(tls_creds_.hash(), secondary_.getReceivedTlsCredsHash());
 }
 
+INSTANTIATE_TEST_SUITE_P(
+    SecondaryRpcTestCases, SecondaryRpcTest,
+    ::testing::Values(std::make_pair(1024 * 3 + 1, true /* new API/messages */),
+                      std::make_pair(1024 * 3 + 1, false /* old messages/sendFirmware, fallback testing */),
```

Comment:
What do you mean by "fallback testing" here? I'm not quite sure what's being covered here. I think this file could use just a bit more documentation about what all is actually being tested here (and what isn't); it's a bit hard to follow.

Reply:
The second instantiation (false) creates an instance of secondary mock that emulates the "old" Secondary, the one that can handle current RPC version between Primary and Secondary (specifically this sendFirmware request/response.) I'll add a comment into a test.